src/hclust.jl

```diff
@@ -531,6 +531,108 @@ function rorder!(hmer::HclustMerges)
     return hmer
 end
 
+
+"""
+    orderleaves_barjoseph!(hmer::HclustMerges, dm::AbstractMatrix)
+
+Given a hierarchical cluster and the distance matrix used to generate it,
+use fast algorithm to determine optimal leaf order minimizing the distance
+between adjacent leaves. This is done using a heuristic where,
+when combining multi-leaf sub branches, only the outermost leaves are
+compared (a maximum of 4 comparisons per intersection).
+
+Sub branches are flipped if necessary to minimize the distance between adjacent
+nodes, and then the combined branches are treated as a block for future
+comparisons.
+
+Based on:
+[Bar-Joseph et. al. "Fast optimal leaf ordering for hierarchical clustering." _Bioinformatics_. (2001)](https://doi.org/10.1093/bioinformatics/17.suppl_1.S22)
+"""
+function orderleaves_barjoseph!(hmer::HclustMerges, dm::AbstractMatrix)
+    order = invperm(hclust_perm(hmer))
+    node_ranges = Tuple{Int,Int}[] # ranges of order array indices occupied by the leaves of each node
+
+    for v in 1:nnodes(hmer)-1
+        vl, vr = hmer.mleft[v], hmer.mright[v]
+
+        (uidx, midx) = node_range(vl, order, node_ranges)
+        (kidx, widx) = node_range(vr, order, node_ranges)
+
+        (u, m) = (order[uidx], order[midx])
+        (k, w) = (order[kidx], order[widx])
+
+        if vl < 0 && vr < 0
+            # Nothing needs to be done
+        elseif vl < 0
+            # check if flipping would reduce distance
+            if dm[m,k] > dm[m,w]
+                reverse!(order, uidx, midx)
+                rotate_merges!(hmer, vr)
+            end
+        elseif vr < 0
+            if dm[k,m] > dm[k,u]
+                reverse!(order, uidx, midx)
+                rotate_merges!(hmer, vl)
+            end
+        elseif vl > 0 && vr > 0
+            # For 2 multi-leaf branches, determine if one or two flips is required
+            # 1 = do not flip
+            # 2 = flip left
+            # 3 = flip right
+            # 4 = flip both
+            flp = argmin((dm[m,k], dm[u,k], dm[m,w], dm[u,w]))
+            if flp == 2 || flp == 4
+                reverse!(order, uidx, midx)
+                rotate_merges!(hmer, vl)
+            end
+            if flp == 3 || flp == 4
+                reverse!(order, kidx, widx)
+                rotate_merges!(hmer, vr)
+            end
+        else
+            error("invalid 'merge' order in Hclust: ($vl, $vr) ")
+        end
+
+        push!(node_ranges, (uidx, widx))
+    end
+    @show order
+end
+
+"""
+    node_range(v::Int, order::Vector{Int}, node_ranges::Vector{Tuple{Int,Int}})
+
```

Comment:
```suggestion
Get the left and right bounds of the range of the `order` array indices occupied by the elements of the `v` node.
```

Actually, we don't need to be more verbose than that, since it's an internal function (we even don't need to have it as a docstring, the comment would have been also fine).

Reply:
Would you prefer a comment? I will definitely not remember how these things work in 3 months without being more verbose

Reply:
> Would you prefer a comment?

Yes, the reason is that maintaining docstring (e.g. syncing arg names and types) is an extra work.

>  I will definitely not remember how these things work in 3 months without being more verbose

Same is true for me :) But then I'd rather add the comments in the method body, since it's an internal function. E.g. `v < 0 # the leaf node`